File: The boolean order/Python/solution_2.py
def solve(s, ops):
    n = len(s)
    D = [[[0, 0] for _ in range(n)] for _ in range(n)]
    for i in range(n):
        D[i][i][s[i] == 't'] = 1
    for l in range(2, n + 1):
        for i in range(n - l + 1):
            j = i + l - 1
            for k in range(i, j):
                for x in [0, 1]:
                    for y in [0, 1]:
                        if ops[k] == '&':
                            D[i][j][x&y] += D[i][k][x] * D[k+1][j][y]
                        elif ops[k] == '|':
                            D[i][j][x|y] += D[i][k][x] * D[k+1][j][y]
                        else:
                            D[i][j][x^y] += D[i][k][x] * D[k+1][j][y]
    return D[0][n-1][1]

File: The boolean order/C++/solution_2.cpp
#include <bits/stdc++.h>
using namespace std;

#define Iter(i, a, b) for (int i = a, _b = b; i < _b; i++)
#define Downto(i, a, b) for (int i = a, _b = b; i >= _b; i--)

long long solve(const string &s, const string &op){
    int n = s.size();
    vector<vector<long long> >f[2];
    Iter(t, 0, 2){
        f[t].resize(n);
        for (auto& a:f[t]) a.resize(n);
    }
    Iter(i, 0, n) f[s[i] == 't'][i][i] = 1;
    #define TRANS(OP) {Iter(t1, 0, 2) Iter(t2, 0, 2) f[t1 OP t2][i][j] += f[t1][i][k] * f[t2][k+1][j];}
    Downto(i, n-1, 0) Iter(j, i+1, n) Iter(k, i, j){
        if (op[k]=='&') TRANS(&);
        if (op[k]=='|') TRANS(|);
        if (op[k]=='^') TRANS(^);
    }
    return f[1][0][n-1];
}